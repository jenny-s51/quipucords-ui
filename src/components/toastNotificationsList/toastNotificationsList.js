import React from 'react';
import PropTypes from 'prop-types';
import { Alert, AlertActionCloseButton, AlertGroup, AlertVariant } from '@patternfly/react-core';
import { connect, reduxTypes, store } from '../../redux';
import helpers from '../../common/helpers';

/**
 * Toast notifications list: operates by allowing mutation of the passed/original toast object from state.
 */
class ToastNotificationsList extends React.Component {
  onHover = toast => {
    store.dispatch({
      type: reduxTypes.toastNotifications.TOAST_PAUSE,
      toast
    });
  };

  onLeave = toast => {
    store.dispatch({
      type: reduxTypes.toastNotifications.TOAST_RESUME,
      toast
    });
  };

  onDismiss = toast => {
    store.dispatch({
      type: reduxTypes.toastNotifications.TOAST_REMOVE,
      toast
    });
  };

  render() {
    const { toasts, timeout } = this.props;

    return (
      <AlertGroup isLiveRegion isToast className="quipucords-toast-notifications__alert-group">
        {toasts?.map(toast => {
          if (!toast.removed) {
            return (
              <Alert
                title={toast.header || toast.message}
                truncateTitle={2}
                timeout={!toast.paused && timeout}
                onTimeout={() => this.onDismiss(toast)}
                variant={toast.alertType}
                actionClose={<AlertActionCloseButton onClose={() => this.onDismiss(toast)} />}
                key={helpers.generateId('key')}
                onMouseEnter={() => this.onHover(toast)}
                onMouseLeave={() => this.onLeave(toast)}
              >
                {(toast.header && toast.message) || ''}
              </Alert>
            );
          }
          return null;
        })}
      </AlertGroup>
    );
  }
}

ToastNotificationsList.propTypes = {
  toasts: PropTypes.arrayOf(
    PropTypes.shape({
      alertType: PropTypes.oneOf([...Object.values(AlertVariant)]),
      header: PropTypes.node,
      message: PropTypes.node,
      removed: PropTypes.bool
    })
  ),
  timeout: PropTypes.number
};

ToastNotificationsList.defaultProps = {
  toasts: [],
  timeout: helpers.TOAST_NOTIFICATIONS_TIMEOUT
};

const mapStateToProps = state => ({ ...state.toastNotifications });

const ConnectedToastNotificationsList = connect(mapStateToProps)(ToastNotificationsList);

export { ConnectedToastNotificationsList as default, ConnectedToastNotificationsList, ToastNotificationsList };
