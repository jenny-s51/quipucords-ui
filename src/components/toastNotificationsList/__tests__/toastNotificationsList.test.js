import React from 'react';
import configureMockStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import { shallow } from 'enzyme';
import { AlertVariant } from '@patternfly/react-core';
import { ConnectedToastNotificationsList, ToastNotificationsList } from '../toastNotificationsList';

describe('ToastNotificationsList Component', () => {
  const generateEmptyStore = () => configureMockStore()({ toastNotifications: {} });

  it('should shallow render a basic component', () => {
    const store = generateEmptyStore();
    const props = {};
    const component = shallow(
      <Provider store={store}>
        <ConnectedToastNotificationsList {...props} />
      </Provider>
    );

    expect(component.find(ConnectedToastNotificationsList)).toMatchSnapshot('basic');
  });

  it('should handle toast variations', () => {
    const props = {
      toasts: [
        {
          alertType: undefined,
          header: undefined,
          message: 'Lorem ipsum',
          removed: undefined,
          paused: undefined
        },
        {
          alertType: AlertVariant.success,
          header: 'Dolor sit',
          message: undefined,
          removed: undefined,
          paused: undefined
        },
        {
          alertType: AlertVariant.danger,
          header: 'Dolor sit',
          message: 'Lorem ipsum',
          removed: undefined,
          paused: undefined
        },
        {
          alertType: AlertVariant.info,
          header: 'REMOVED, Dolor sit',
          message: undefined,
          removed: true,
          paused: undefined
        },
        {
          alertType: AlertVariant.info,
          header: 'PAUSED, Dolor sit',
          message: undefined,
          removed: undefined,
          paused: true
        }
      ]
    };
    const component = shallow(<ToastNotificationsList {...props} />);

    expect(component).toMatchSnapshot('variations');
  });
});
