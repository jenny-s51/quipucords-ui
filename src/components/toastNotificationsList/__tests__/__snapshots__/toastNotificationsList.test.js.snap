// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToastNotificationsList Component should handle toast variations: variations 1`] = `
<AlertGroup
  className="quipucords-toast-notifications__alert-group"
  isLiveRegion={true}
  isToast={true}
>
  <Alert
    actionClose={
      <AlertActionCloseButton
        onClose={[Function]}
      />
    }
    key="key-"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onTimeout={[Function]}
    timeout={8000}
    title="Lorem ipsum"
    truncateTitle={2}
  />
  <Alert
    actionClose={
      <AlertActionCloseButton
        onClose={[Function]}
      />
    }
    key="key-"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onTimeout={[Function]}
    timeout={8000}
    title="Dolor sit"
    truncateTitle={2}
    variant="success"
  />
  <Alert
    actionClose={
      <AlertActionCloseButton
        onClose={[Function]}
      />
    }
    key="key-"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onTimeout={[Function]}
    timeout={8000}
    title="Dolor sit"
    truncateTitle={2}
    variant="danger"
  >
    Lorem ipsum
  </Alert>
  <Alert
    actionClose={
      <AlertActionCloseButton
        onClose={[Function]}
      />
    }
    key="key-"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onTimeout={[Function]}
    timeout={false}
    title="PAUSED, Dolor sit"
    truncateTitle={2}
    variant="info"
  />
</AlertGroup>
`;

exports[`ToastNotificationsList Component should shallow render a basic component: basic 1`] = `<Connect(ToastNotificationsList) />`;
